### PR TITLE
Introduced Custom ProtoDataStore to ensure type-safety

### PR DESCRIPTION
**Pull request by 07jasjeet**
This implementation results in flexibility and easier addition of preferences. Any new `DataStore`s can use the new `ProtoDataStore` class to ensure type-safety.

### CODE REVIEW

File: app/src/main/java/org/listenbrainz/android/repository/listenservicemanager/ListenServiceManagerImpl.kt [hunk 1]
```diff
@@ -40,7 +41,7 @@ class ListenServiceManagerImpl @Inject constructor(
     /** Used to avoid repetitive submissions.*/
     private var lastNotificationPostTs = System.currentTimeMillis()
     private lateinit var whitelist: List<String>
-    private var isScrobblingAllowed: Boolean = true
+    private var isListeningAllowed: AtomicBoolean = AtomicBoolean(true)
```
Comment (akshaaatt) [inline app/src/main/java/org/listenbrainz/android/repository/listenservicemanager/ListenServiceManagerImpl.kt:44]: I see you are trying to make this thread safe, but why was this necessary? Can you please link the issue this solves?
Also, why is it being solved in this PR. Totally irrelevant

File: app/src/main/java/org/listenbrainz/android/util/datastore/ProtoDataStore.kt [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+import kotlinx.coroutines.flow.catch
+import kotlinx.coroutines.flow.firstOrNull
+import kotlinx.coroutines.flow.map
+import org.listenbrainz.android.model.ResponseError
+import org.listenbrainz.android.util.Resource
+import org.listenbrainz.android.util.datastore.Preference.Companion.ComplexPreference
+import org.listenbrainz.android.util.datastore.Preference.Companion.PrimitivePreference
+import java.io.IOException
+
+abstract class ProtoDataStore(private val dataStore: DataStore<Preferences>) {
```
Comment (akshaaatt) [inline app/src/main/java/org/listenbrainz/android/util/datastore/ProtoDataStore.kt:16]: @07jasjeet  I see you have made a library to do this. Can we import these from there instead of adding chunks of files to our codebase?

**Review by akshaaatt — CHANGES_REQUESTED** [review]

File: app/src/main/java/org/listenbrainz/android/util/datastore/ProtoDataStore.kt (hunk 2 above)
Comment (07jasjeet) [reply app/src/main/java/org/listenbrainz/android/util/datastore/ProtoDataStore.kt:16]: Yes. On it.

File: app/src/main/java/org/listenbrainz/android/repository/listenservicemanager/ListenServiceManagerImpl.kt (hunk 1 above)
Comment (07jasjeet) [reply app/src/main/java/org/listenbrainz/android/repository/listenservicemanager/ListenServiceManagerImpl.kt:44]: Replied on IRC.

File: app/src/main/java/org/listenbrainz/android/util/datastore/ProtoDataStore.kt (hunk 2 above)
Comment (akshaaatt) [reply app/src/main/java/org/listenbrainz/android/util/datastore/ProtoDataStore.kt:16]: @07jasjeet any update on this?

Comment (07jasjeet) [reply app/src/main/java/org/listenbrainz/android/util/datastore/ProtoDataStore.kt:16]: Will update by weekend.

**Comment by akshaaatt** [comment]
I think it is better to close this PR and open a fresh PR for the same
